libantworld/agent.h

```diff
@@ -21,67 +35,279 @@ using namespace units::literals;
 //----------------------------------------------------------------------------
 //! A simple agent with a position and a panoramic view of the current AntWorld
 class AntAgent
-  : public Video::OpenGL
+  : public Robots::Robot
+  , public Video::OpenGL
 {
-    using degree_t = units::angle::degree_t;
     using meter_t = units::length::meter_t;
+    using meters_per_second_t = units::velocity::meters_per_second_t;
+    using degree_t = units::angle::degree_t;
+    using radians_per_second_t = units::angular_velocity::radians_per_second_t;
 
 public:
-    AntAgent(GLFWwindow *window, Renderer &renderer, GLsizei readWidth, GLsizei readHeight)
-      : Video::OpenGL(0, 0, readWidth, readHeight)
+    static constexpr meters_per_second_t DefaultVelocity = 300_mm / 1_s;
+    static constexpr radians_per_second_t DefaultTurnSpeed = 200_deg_per_s;
+
+    AntAgent(GLFWwindow *window, Renderer &renderer,
+             const cv::Size &renderSize,
+             meters_per_second_t velocity = DefaultVelocity, radians_per_second_t turnSpeed = DefaultTurnSpeed)
+      : Video::OpenGL(renderSize)
+      , m_Velocity(velocity)
+      , m_TurnSpeed(turnSpeed)
       , m_Renderer(renderer)
       , m_Window(window)
     {}
 
-    template<class LengthUnit = meter_t>
-    Vector3<LengthUnit> getPosition() const
+    void addJoystick(HID::Joystick &joystick, float deadZone = 0.25f)
     {
-        return convertUnitArray<LengthUnit>(m_Position);
+        joystick.addHandler(
+                [this, deadZone](HID::JAxis axis, float value) {
+                    switch (axis) {
+                    case HID::JAxis::LeftStickVertical:
+                        m_JoystickY = value;
+                        break;
+                    case HID::JAxis::LeftStickHorizontal:
+                        m_JoystickX = value;
+                        break;
+                    default:
+                        return false;
+                    }
+
+                    if (fabs(m_JoystickY) > deadZone) {
+                        moveForward(-m_JoystickY);
+                    } else if (fabs(m_JoystickX) > deadZone) {
+                        turnOnTheSpot(m_JoystickX);
+                    } else {
+                        stopMoving();
+                    }
+                    return true;
+                });
     }
 
-    template<class AngleUnit = degree_t>
-    Vector3<AngleUnit> getAttitude() const
+    template<typename LengthUnit = meter_t>
+    Vector3<LengthUnit> getPosition()
     {
-        return convertUnitArray<AngleUnit>(m_Attitude);
+        std::lock_guard<std::mutex> guard(m_PoseMutex);
```

Comment:
maybe make ``m_PostMutex`` mutable so you can make this API const

Reply:
...although it turns out it calls ``updatePose()`` which sets some other vars. Although it is a little odd to have a non-const getter, I think it does actually make sense here (because the pose is being updated based on elapsed time every time it's called).

Reply:
That seems like bad getter behaviour to me! I feel it would be nicer if``updatePost`` just get called once a frame from ``update`` and ``getPosition`` returns that frame's state? I've definitely lost track of what this does since 20th December :smile: so that may make no sense at all

Reply:
Haha, yes, sorry for leaving it for so long!

That does make sense. I'm thinking actually that having the ``Stopwatch`` inside this class is a mistake. Might be better to let the user call ``updatePose()`` and pass in the elapsed time as an arg, so we don't always have to run the simulation in real time. Whaddya think?

Reply:
Agreed - that sounds much better